### PR TITLE
T02: Feature- Admin can create a new job

### PR DESCRIPTION
**Pull request by shinbe69**
Create an API help admin user can post a new job.
Link deployed server: [https://onboarding-job-board-api.herokuapp.com].
How to test API:
1. Use Postman and add the above server link in the text field called "Enter request URL"
2. Login to the server by using "/api/sessions" API, then switch to tab Body below and type in the credential information in JSON format.  Remember to choose POST method for the request and click Send
3. Information about the login session should be appear in the result field in the bottom. These information will be used for "/api/job" API.
4. In "/api/job" API, we create an input is information about the job in JSON format and click Send.
5. If everything work fine, all information about that job will be created in database.

Before T02 API was created:
before

After:
after

### CODE REVIEW

File: src/persistence/users.js [hunk 1]
```diff
@@ -28,6 +28,6 @@ module.exports = {
     const {rows} = await db.query(sql`
     SELECT * FROM users WHERE email=${email} LIMIT 1;
     `);
-    return rows[0];
+    return rows;
```
Comment (tronghiendev) [inline src/persistence/users.js]: Why do we need this change?

File: config.js [hunk 2]
```diff
@@ -1,3 +1,3 @@
 module.exports = {
   SECRET: process.env.SECRET || 'faba-secret-key'
-};
+};
```
Comment (tronghiendev) [inline config.js]: Please don't commit the changes that unnecessary for the task

Maybe this is caused by ide's code format, please turn off or don't commit them

File: package.json [hunk 3]
```diff
@@ -20,6 +20,7 @@
   "license": "MIT",
   "dependencies": {
     "bcrypt": "^5.0.0",
+    "cookie-parser": "^1.4.5",
```
Comment (tronghiendev) [inline package.json]: What is this package used for?

If not, please remove it

File: src/api/index.js [hunk 4]
```diff
@@ -5,11 +5,19 @@ const router = new Router();
...
 const user = require('./user');
 const session = require('./session');
+const createJob = require('./createJob');
+// const viewJob = require('./viewJob');
+// const updateJob = require('./updateJob');
+// const deleteJob = require('./deleteJob');
 
 router.use('/api/health', (request, response) => {
-  response.status(200).json();
+  response.status(400).json();
```
Comment (tronghiendev) [inline src/api/index.js]: Why need to change the status code here?

File: src/persistence/createJob.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+const sql = require('sql-template-strings');
```
Comment (tronghiendev) [inline src/persistence/createJob.js]: I think this file should be job.js and it has methods like create, update, delete,...

File: src/persistence/createJob.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+const sql = require('sql-template-strings');
+const {v4: uuidv4} = require('uuid');
+const bcrypt = require('bcrypt');
+const db = require('./db');
+
+module.exports = {
+    async create(title, salary, description, tags, company, logoURL) {
```
Comment (tronghiendev) [inline src/persistence/createJob.js]: Try to use object here, e.g: `{title, salary, description, tags, company, logoURL}`

It will be very easier if we want to add a field

File: src/persistence/postgres-state-storage.js [hunk 7]
```diff
@@ -49,4 +49,4 @@ const postgresStateStorage = {
 
 module.exports = Object.assign(() => {
   return postgresStateStorage;
-}, postgresStateStorage);
+}, postgresStateStorage);
```
Comment (tronghiendev) [inline src/persistence/postgres-state-storage.js]: Same here

File: package-lock.json [hunk 8]
```diff
@@ -1361,6 +1361,15 @@
       "resolved": "https://registry.npmjs.org/cookie/-/cookie-0.4.0.tgz",
       "integrity": "sha512-+Hp8fLp57wnUSt0tY0tHEXh4voZRDnoIrZPqlo3DPiI4y9lwg/jqx+1Om94/W6ZaPDOUbnjOt/99w66zk+l1Xg=="
     },
+    "cookie-parser": {
```
Comment (tronghiendev) [inline package-lock.json]: If you remove this package in package.json, please also remove here

File: src/api/createJob.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
+const {Router} = require('express');
```
Comment (tronghiendev) [inline src/api/createJob.js]: Same here, the API file should be job.js

**Review by tronghiendev — COMMENTED** [review]
Hi @shinbe69 ,

Please target this PR into `<yourname>/job_board_api` branch and edit the commit message to make sense

Try to write the unit test and deploy to a server 

Also, I have some reviews, please check

File: src/migrations/1602166719410-authentication.js [hunk 10]
```diff
@@ -9,6 +9,16 @@ module.exports.up = async function (next) {
     email text UNIQUE,
     password text
   );
+  CREATE TABLE IF NOT EXISTS jobs (
```
Comment (tronghiendev) [inline src/migrations/1602166719410-authentication.js]: Should create a new migration for this

File: src/persistence/users.js (hunk 1 above)
Comment (shinbe69) [reply src/persistence/users.js]: Actually, it's just a typing mistake, I fixed it.

File: config.js (hunk 2 above)
Comment (shinbe69) [reply config.js]: This is also a typing mistake, I fixed it as well.

File: src/persistence/createJob.js (hunk 5 above)
Comment (shinbe69) [reply src/persistence/createJob.js]: Updated

File: src/persistence/createJob.js (hunk 6 above)
Comment (shinbe69) [reply src/persistence/createJob.js]: Updated

File: package.json (hunk 3 above)
Comment (shinbe69) [reply package.json]: I thought we need to save token information into a cookie and send back to the user when signing in, but I found that we don't need it anymore, so I removed it.

File: src/api/index.js (hunk 4 above)
Comment (shinbe69) [reply src/api/index.js]: This is a mistake, it occurred when I try to view the code and change it to see what happened. I fixed it.

File: src/persistence/postgres-state-storage.js (hunk 7 above)
Comment (shinbe69) [reply src/persistence/postgres-state-storage.js]: Updated.

File: package-lock.json (hunk 8 above)
Comment (shinbe69) [reply package-lock.json]: Updated.

File: src/api/createJob.js (hunk 9 above)
Comment (shinbe69) [reply src/api/createJob.js]: Updated.

File: src/migrations/1602166719410-authentication.js (hunk 10 above)
Comment (shinbe69) [reply src/migrations/1602166719410-authentication.js]: New migration for jobs table was created.

File: src/api/index.js [hunk 11]
```diff
@@ -5,11 +5,13 @@ const router = new Router();
...
 const user = require('./user');
 const session = require('./session');
+const job = require('./job');
 
 router.use('/api/health', (request, response) => {
   response.status(200).json();
 });
 router.use('/api/users', user);
 router.use('/api/sessions', session);
+router.use('/api/job', job);
```
Comment (leofaba) [inline src/api/index.js]: We should use the plural for the resource names. Please take a look at the onboarding document https://www.notion.so/mrleo5/REST-API-ac5d52964c3c4ef59be9f88058221d5c for details.

File: src/api/job.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
+const {Router} = require('express');
+const Job = require('../persistence/job');
+
+const router = new Router();
+
+// Admin Create a new job with post method
```
Comment (leofaba) [inline src/api/job.js]: **The rule is to try to NOT comment in your code.**
- If your code is clear, don't need comment
- If your code is complicated that you think you have to explain it, refactor the complicated code instead.

File: src/api/job.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+      logoURL
+    } = request.body;
+    if (
+      !title ||
+      !salaryrange ||
+      !description ||
+      !tags ||
+      !company ||
+      !logoURL
+    ) {
```
Comment (leofaba) [inline src/api/job.js:23]: The condition format is hard to follow if we break it into multiple lines. Should we format it to one line instead?

File: src/api/job.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+      return response
+        .status(400)
+        .json({message: 'All information must be provided!'});
+    }
+
+    const para = {
```
Comment (leofaba) [inline src/api/job.js]: Name variable clearly and abbreviation restrictions. We write the code for the team, not only for us. Let's think about how other members of our team read the code and understand it easily. For example, `parameters` in this case, not `para`, `pa`, or `p`.

File: src/api/job.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+const Job = require('../persistence/job');
+
+const router = new Router();
+
+// Admin Create a new job with post method
+router.post('/', async (request, response) => {
+  try {
+    const {
+      title,
+      salaryrange,
```
Comment (leofaba) [inline src/api/job.js]: If the variable name contains more than 1 word, we should use the camel case format.

File: src/migrations/1603170477620-job.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+  const client = await db.connect();
+  await client.query(`CREATE TABLE IF NOT EXISTS jobs (
+      id uuid PRIMARY KEY,
+      title text,
+      salaryRange text,
+      description text,
+      createAt timestamp DEFAULT CURRENT_TIMESTAMP,
+      tags text[],
+      company text,
+      logoURL text DEFAULT 'https://unsplash.com/photos/g2E2NQ5SWSU'
```
Comment (leofaba) [inline src/migrations/1603170477620-job.js]: Just a small note, I don't think the image link works. It will open the whole unsplash website instead of the image only. You can try it https://images.unsplash.com/photo-1593642632559-0c6d3fc62b89?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80

File: src/persistence/job.js [hunk 17]
```diff
@@ -0,0 +1,32 @@
+const sql = require('sql-template-strings');
+const {v4: uuidv4} = require('uuid');
+const db = require('./db');
+
+module.exports = {
+  // Method create() will get all information about the job without id and create a new job with those data
```
Comment (leofaba) [inline src/persistence/job.js]: The same comment above. The rule is comment restriction.

File: test/job.js [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+          logoURL: 'https://unsplash.com/photos/g2E2NQ5SWSU'
+        };
+        chai
+          .request(app())
+          .post('/api/job')
+          .set('Authorization', 'Bearer ' + accessToken)
+          .send(job)
+          .end((err, response) => {
+            response.should.have.status(201);
+            // Response.body.should.be.a('json');
```
Comment (tronghiendev) [inline test/job.js]: Please remove the commented code

File: src/api/index.js (hunk 11 above)
Comment (shinbe69) [reply src/api/index.js]: Updated.

File: src/api/job.js (hunk 12 above)
Comment (shinbe69) [reply src/api/job.js]: I fixed it.

File: src/api/job.js (hunk 13 above)
Comment (shinbe69) [reply src/api/job.js:23]: This one is the result of the format "xo prettier" when I run the unit test. It doesn't allow write those conditions in one line. I'm trying to figure out why.

File: src/api/job.js (hunk 14 above)
Comment (shinbe69) [reply src/api/job.js]: I understood, thank you.

File: src/api/job.js (hunk 15 above)
Comment (shinbe69) [reply src/api/job.js]: Updated.

File: src/migrations/1603170477620-job.js (hunk 16 above)
Comment (shinbe69) [reply src/migrations/1603170477620-job.js]: I used the direct link of the photo instead. Thank you.

File: src/persistence/job.js (hunk 17 above)
Comment (shinbe69) [reply src/persistence/job.js]: Updated.

File: test/job.js (hunk 18 above)
Comment (shinbe69) [reply test/job.js]: Updated.

File: src/api/job.js (hunk 13 above)
Comment (leofaba) [reply src/api/job.js:23]: Ah okey. If it's the default settings of `xo prettier`, we can use it.

File: src/persistence/job.js [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+        VALUES (${uuidv4()}, ${jobParameters.title}, ${
+        jobParameters.salaryRange
+      }, ${jobParameters.description}, current_date, ${jobParameters.tags}, ${
+        jobParameters.company
+      }, ${jobParameters.logoURL})
+        RETURNING id, title, tags;
+      `);
+      return rows[0];
+    } catch (error) {
+      if (error.constraint === 'users_email_key') {
```
Comment (leofaba) [inline src/persistence/job.js]: Why do we check this?

Comment (shinbe69) [reply src/persistence/job.js]: This one is a mistake because I copied it from persistence/users.js and modified without checking.

Comment (leofaba) [reply src/persistence/job.js]: Can you remove it? The constraint `users_email_key` belongs to `users` table which has unique email. For the `jobs` table, we don't have it right?

And please don't copy and paste the code that you don't understand.

**Review by leofaba — CHANGES_REQUESTED** [review]
